pep-0727.rst

```diff
@@ -0,0 +1,516 @@
+editors and other tools could implement support for it.
+
+This doesn't deprecate current usage of docstrings, it's transparent to common
+developers (library users), and it's only opt-in for library authors that would
+like to adopt it.
+
+
+Specification
+=============
+
+
+``typing.doc``
+--------------
+
+The main proposal is to have a new function ``doc()`` in the ``typing`` module.
+Even though this is not strictly related to the type annotations, it's expected
+to go in ``Annotated`` type annotations, and to interact with type annotations.
+
+There's also the particular benefit that it could be implemented in the
+``typing_extensions`` package to have support for older versions of Python and
+early adopters of this proposal.
+
+This ``doc()`` function would receive one single parameter ``documentation`` with
+a documentation string.
+
+This string could be a multi-line string, in which case, when extracted by tools,
+should be interpreted cleaning up indentation as if using ``inspect.cleandoc()``,
+the same procedure used for docstrings.
+
+This string could probably contain markup, like Markdown or reST. As that could
+be highly debated, that decision is left for a future proposal, to focus here
+on the main functionality.
+
+This specification targets static analysis tools and editors, and as such, the
+value passed to ``doc()`` should allow static evaluation and analysis. If a
+developer passes as the value something that requires runtime execution
+(e.g. a function call) the behavior of static analysis tools is unspecified
+and they could omit it from their process and results. For static analysis
+tools to be conformant with this specification they need only to support
+statically accessible values.
+
+An example documenting the attributes of a class, or in this case, the keys
+of a ``TypedDict``, could look like this:
+
+.. code-block::
+
+    from typing import Annotated, TypedDict, NotRequired, doc
+
+
+    class User(TypedDict):
+        firstname: Annotated[str, doc("The user's first name")]
+        lastname: Annotated[str, doc("The user's last name")]
+
+
+An example documenting the parameters of a function could look like this:
+
+.. code-block::
+
+    from typing import Annotated, doc
+
+
+    def create_user(
+        lastname: Annotated[str, doc("The **last name** of the newly created user")],
+        firstname: Annotated[str | None, doc("The user's **first name**")] = None,
+    ) -> Annotated[User, doc("The created user after saving in the database")]:
+        """
+        Create a new user in the system, it needs the database connection to be already
+        initialized.
+        """
+        pass
+
+
+The return of the ``doc()`` function is an instance of a class that can be checked
+and used at runtime, defined similar to:
+
+.. code-block::
+
+    class DocInfo:
+        def __init__(self, documentation: str):
+            self.documentation = documentation
```

Comment:
Why not just use ``DocInfo(...)`` -- the need for two distinct names is unexplained.